rfcs/0013-configurable-and-composable-storage.md

```diff
@@ -0,0 +1,600 @@
+
+Thus, the implementations of storage types are essentially recursive compositions of their contained storage types. The `Value` and the `Config` associated types will, in general, be constructed based on the `Value` and `Config` types of the contained storage types. Similarly, the implementations of the `internal_get_config()` and `init()` functions will recursively call those same methods of the contained storage types.
+
+To get the feeling for these straightforward and essentially short and compact implementations of storage types, see the atomic and compound storage types implemented in [`sway-libs`](../files/0013-configurable-and-composable-storage/sway-libs/) and the user-defined [`StoragePair`](../files/0013-configurable-and-composable-storage/user-defined-libs/storage_pair.sw).
+
+### Defining storage traits
+
+[defining-storage-traits]: #defining-storage-traits
+
+The clear definition of the `Storage` trait allows defining storage traits. A storage trait is a trait that has `Storage` as a supertrait. The STD will come with (at least) two storage traits, `DeepReadStorage` and `DeepClearStorage` both defined in the [`storage.sw`](../files/0013-configurable-and-composable-storage/sway-libs/storage.sw).
+
+The implementations of storage traits will, similar to the implementations of `Storage`, be straightforward and based on recursive calls to the implementations of the same storage traits on the contained storage types.
+
+To get the feeling for these straightforward and essentially short and compact implementations of storage traits, see the implementations of `DeepReadStorage` and `DeepClearStorage` for [`StorageVec`](../files/0013-configurable-and-composable-storage/sway-libs/storage/storage_vec.sw) and [`StoragePair`](../files/0013-configurable-and-composable-storage/user-defined-libs/storage_pair.sw).
+
+By convention, all storage trait will have the suffix `Storage`.
+
+## Internal API and the `__state` intrinsics
+
+[internal-api-and-the-state-intrinsics]: #internal-api-and-the-state-intrinsics
+
+The STD will still provide the low-level storage `read`, `write`, and `clear` functions. However, unlike now when they are actually sometimes needed in contract code, e.g., to store arrays, in the new storage implementation there should never be a need to use them in contracts. Instead, the atomic `StorageBox` and `StorageEncodedBox` should be used. They provide the same low-level functionality while offering safe storage access.
+
+The low-level API will be used only when implementing storage types, and even in those cases not always. Thus, the proposal is to move them to the module named [`internal`](../files/0013-configurable-and-composable-storage/sway-libs/storage/internal.sw) to emphasize that they are, similar to `Storage::internal_` functions, meant to be used only when developing custom storage types.
```

Comment:
Direct programmatic manipulation of storage should not be encouraged but it should not be this discouraged either. There are cases where it's a better more straightforward pattern to use to implement what you want.

I do think having it in a separate module is fine, but I don't like the implication that it is "internal", it's merly raw or low level. But it's perfectly valid for the end user and not specific to storage type implementors.

Reply:
Agree!

Encouraging and/or discouraging usage is a separate topic, similar to the one of using raw pointers in contract code. It is to be reflected in the way how we convey and document the library/language features, patterns, etc. But in both cases, even discouraging the usage does not justify the name "internal".

The current module name is `storage_api` which we can keep. But, I am more for `raw_api` or some other classifier which will better convay that this is the low-level API, rarely needed, if at all, in the contract code.

@bitzoic Do you have a view or proposal here?